{
  "metadata" : {
    "id" : "e8f2b8c5-813f-46e1-a81d-69d02380d726",
    "name" : "lags_animals",
    "user_save_timestamp" : "2018-02-24T23:14:56.713Z",
    "auto_save_timestamp" : "1970-01-01T00:00:00.000Z",
    "language_info" : {
      "name" : "scala",
      "file_extension" : "scala",
      "codemirror_mode" : "text/x-scala"
    },
    "trusted" : true,
    "sparkNotebook" : null,
    "customLocalRepo" : "/root/.coursier/cache/v1",
    "customRepos" : [ "comp-bio-aging % default % https://dl.bintray.com/comp-bio-aging/main/", "denigma % default % https://dl.bintray.com/denigma/denigma-releases/", "releases % default % https://oss.sonatype.org/content/repositories/releases", "snapshots % default % https://oss.sonatype.org/content/repositories/snapshots/" ],
    "customDeps" : [ "org.bdgenomics.adam %% adam-core-spark2 % 0.23.0", "comp.bio.aging %% adam-playground % 0.0.9", "com.crealytics %% spark-excel % 0.9.12", "org.typelevel %% frameless-cats    % 0.5.0", "org.typelevel %% frameless-dataset   % 0.5.0", "com.github.pathikrit  %% better-files  % 3.4.0", "org.apache.hadoop % hadoop-azure % 2.7.5" ],
    "customImports" : null,
    "customArgs" : null,
    "customSparkConf" : {
      "spark.master" : "spark://spark-master:7077",
      "spark.executor.cores" : "8",
      "spark.executor.memory" : "15G",
      "spark.serializer" : "org.apache.spark.serializer.KryoSerializer",
      "spark.kryo.registrator" : "org.bdgenomics.adam.serialization.ADAMKryoRegistrator",
      "spark.kryo.referenceTracking" : "true"
    },
    "customVars" : null
  },
  "cells" : [ {
    "metadata" : {
      "id" : "D907F31AC42F48BE98A8642CDC15C9E4"
    },
    "cell_type" : "markdown",
    "source" : "Longevity Association Genes comparison notebook\n------------------------------------------------------\nThis script takes transcript expressions (in TPM) from different species as well as information about homology and makes a comparison table. "
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "99C31FAA4900457583349F96F5418FA8"
    },
    "cell_type" : "code",
    "source" : [ "import org.apache.spark.sql.{DataFrame, Encoders, SparkSession}\n", "import org.apache.spark.sql.types.StructType\n", "import org.bdgenomics.adam.rdd.ADAMContext._\n", "import org.bdgenomics.adam.rdd.ADAMContextExtensions._\n", "import scala.reflect.runtime.universe._\n", "import comp.bio.aging.playground.extras.uniprot._\n", "import scala.collection.JavaConversions._\n", "import org.apache.spark.storage.StorageLevel" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "import org.apache.spark.sql.{DataFrame, Encoders, SparkSession}\nimport org.apache.spark.sql.types.StructType\nimport org.bdgenomics.adam.rdd.ADAMContext._\nimport org.bdgenomics.adam.rdd.ADAMContextExtensions._\nimport scala.reflect.runtime.universe._\nimport comp.bio.aging.playground.extras.uniprot._\nimport scala.collection.JavaConversions._\nimport org.apache.spark.storage.StorageLevel\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 1,
      "time" : "Took: 4.378s, at 2018-05-03 16:07"
    } ]
  }, {
    "metadata" : {
      "id" : "B1EDF520D38A44B1801BD9E4C4BF1181"
    },
    "cell_type" : "markdown",
    "source" : "Few comments about filesystems\n----------------------------------\nAny Hadoop-compartible filesystem should work with this notebook, including HDFS. \nWe used Azure Blob storage and had to write several \nAzure specific lines here"
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "868CD4FCD5D847628E48E11240DF07CA"
    },
    "cell_type" : "code",
    "source" : [ "def sparkHadoopConf(sc: SparkContext, acountName: String, accountKey: String) = {\n", "  sc.hadoopConfiguration.set(\"fs.azure\", \"org.apache.hadoop.fs.azure.NativeAzureFileSystem\")\n", "  sc.hadoopConfiguration.set(\"fs.azure.account.key.\" + acountName + \".blob.core.windows.net\", accountKey)\n", "  sc\n", "}" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "sparkHadoopConf: (sc: org.apache.spark.SparkContext, acountName: String, accountKey: String)org.apache.spark.SparkContext\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 3,
      "time" : "Took: 1.603s, at 2018-05-03 16:09"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "B8D498D3235F45C982B76EE38D69408F"
    },
    "cell_type" : "code",
    "source" : [ "def azurize(container: String, accountName: String, blobFile: String): String = \"wasbs://\"+container+\"@\"+accountName+\".blob.core.windows.net/\"+blobFile \n", "\n", "def writeText2Azure[T]( rdd: RDD[T], container: String, accountName: String, blobFile: String ): String =\n", "{\n", "  val url = azurize(container, accountName, blobFile)\n", "  rdd.saveAsTextFile(url)\n", "  url\n", "}\n", "\n", "def writeTsv2Azure( df: DataFrame, container: String, accountName: String, blobFile: String ): String =\n", "{\n", "  val url = azurize(container, accountName, blobFile)\n", "  df.write.option(\"sep\",\"\\t\").option(\"header\",\"true\").csv(url)\n", "  url\n", "}\n" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "azurize: (container: String, accountName: String, blobFile: String)String\nwriteText2Azure: [T](rdd: org.apache.spark.rdd.RDD[T], container: String, accountName: String, blobFile: String)String\nwriteTsv2Azure: (df: org.apache.spark.sql.DataFrame, container: String, accountName: String, blobFile: String)String\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 1,
      "time" : "Took: 0.851s, at 2018-05-03 19:28"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : true,
      "id" : "AB1D5FF9D39A493A8AE9B3870DCE3311"
    },
    "cell_type" : "code",
    "source" : [ "//val connString = \"put your connection string here\"\n", "//val key = \"put your connection key here\"\n", "\n", "\n", "val connString = \"DefaultEndpointsProtocol=https;AccountName=pipelines1;AccountKey=;EndpointSuffix=core.windows.net\"\n", "val key = \"\"\n" ],
    "outputs" : [ ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "55721EE2DBA64EA5A6846602DA5CC36C"
    },
    "cell_type" : "code",
    "source" : [ "val account = \"pipelines1\"\n", "def az(path: String): String = azurize(\"storage\", account, path)\n" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "connString: String = DefaultEndpointsProtocol=https;AccountName=pipelines1;AccountKey=;EndpointSuffix=core.windows.net\naccount: String = pipelines1\nkey: String = \naz: (path: String)String\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 2,
      "time" : "Took: 0.767s, at 2018-05-03 19:28"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "C1D04970F17A4BFCAD8A646F50627914"
    },
    "cell_type" : "code",
    "source" : [ "sparkHadoopConf(sparkContext, account, key)\n", "  \n", "val spark = SparkSession\n", "  .builder()\n", "  .appName(\"lags\")\n", "  .getOrCreate()" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "spark: org.apache.spark.sql.SparkSession = org.apache.spark.sql.SparkSession@3f724989\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 3,
      "time" : "Took: 0.864s, at 2018-05-03 19:28"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "19D8895778004E04A98F1BEC66F27322"
    },
    "cell_type" : "code",
    "source" : [ "def loadTranscripts(species: String) = spark.readTSV(az(s\"expressions/transcripts/${species}_transcripts_all.tab\"), true)" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "loadTranscripts: (species: String)org.apache.spark.sql.DataFrame\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 4,
      "time" : "Took: 0.846s, at 2018-05-03 19:28"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "556F5C964FA442148D501A24764B5B72"
    },
    "cell_type" : "code",
    "source" : [ "val quant_base = az(\"quant\")\n", "\n", "val bowhead_liver_path = quant_base + \"/bowhead/liver/transcripts_quant/quant.sf\"\n", "val gray_whale_liver_path = quant_base +\"/gray_whale/liver/transcripts_quant/quant.sf\"\n", "val naked_mole_rat_liver_path = quant_base +\"/NMR/liver/transcripts_quant/quant.sf\"\n", "val minke_whale_liver_path = quant_base +\"/minke_whale/liver/transcripts_quant/quant.sf\"\n", "val naked_mole_rat_ensembl_liver_path = quant_base + \"/NMR_ensembl_female/liver/transcripts_quant/quant.sf\"\n", "val cow_liver_path = quant_base + \"/cow/liver/GSM1020724/quant.sf\"\n", "val human_liver_path = quant_base + \"/human/liver/GSM1698568/quant.sf\"\n", "val mouse_liver_path = quant_base +\"/mouse/liver/GSM1400574/quant.sf\"\n" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "quant_base: String = wasbs://storage@pipelines1.blob.core.windows.net/quant\nbowhead_liver_path: String = wasbs://storage@pipelines1.blob.core.windows.net/quant/bowhead/liver/transcripts_quant/quant.sf\ngray_whale_liver_path: String = wasbs://storage@pipelines1.blob.core.windows.net/quant/gray_whale/liver/transcripts_quant/quant.sf\nnaked_mole_rat_liver_path: String = wasbs://storage@pipelines1.blob.core.windows.net/quant/NMR/liver/transcripts_quant/quant.sf\nminke_whale_liver_path: String = wasbs://storage@pipelines1.blob.core.windows.net/quant/minke_whale/liver/transcripts_quant/quant.sf\nnaked_mole_rat_ensembl_liver_path: String = wasbs://storage@pipelines1.blob.core.windows.net/quant/NMR_ensembl_female/liver/transcripts_quant/quant.sf\ncow_liver_path: String = wasbs://storage@pipelines1.b..."
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 6,
      "time" : "Took: 0.853s, at 2018-05-03 19:29"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "7F91936BCE0147CA835A64D8492BB811"
    },
    "cell_type" : "code",
    "source" : [ "val bowhead_kidney_path = quant_base + \"/bowhead/kidney/transcripts_quant/quant.sf\"\n", "val gray_whale_kidney_path = quant_base +\"/gray_whale/kidney/transcripts_quant/quant.sf\"\n", "val naked_mole_rat_kidney_path = quant_base +\"/NMR/kidney/transcripts_quant/quant.sf\"\n", "val minke_whale_kidney_path = quant_base +\"/minke_whale/kidney/transcripts_quant/quant.sf\"\n", "val naked_mole_rat_ensembl_kidney_path = quant_base + \"/NMR_ensembl_female/kidney/transcripts_quant/quant.sf\"\n", "val cow_kidney_path = quant_base + \"/cow/kidney/GSM1020723/quant.sf\"\n", "val human_kidney_path = quant_base + \"/human/kidney/GSM1698570/quant.sf\"\n", "val mouse_kidney_path= quant_base + \"/mouse/kidney/GSM2195188/quant.sf\" \n" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "bowhead_kidney_path: String = wasbs://storage@pipelines1.blob.core.windows.net/quant/bowhead/kidney/transcripts_quant/quant.sf\ngray_whale_kidney_path: String = wasbs://storage@pipelines1.blob.core.windows.net/quant/gray_whale/kidney/transcripts_quant/quant.sf\nnaked_mole_rat_kidney_path: String = wasbs://storage@pipelines1.blob.core.windows.net/quant/NMR/kidney/transcripts_quant/quant.sf\nminke_whale_kidney_path: String = wasbs://storage@pipelines1.blob.core.windows.net/quant/minke_whale/kidney/transcripts_quant/quant.sf\nnaked_mole_rat_ensembl_kidney_path: String = wasbs://storage@pipelines1.blob.core.windows.net/quant/NMR_ensembl_female/kidney/transcripts_quant/quant.sf\ncow_kidney_path: String = wasbs://storage@pipelines1.blob.core.windows.net/quant/cow/kidney/GSM1020723/quant.sf\nhuman_k..."
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 8,
      "time" : "Took: 0.794s, at 2018-05-03 19:29"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "DA8E2A83C65A42B08A2C7459F9E84A61"
    },
    "cell_type" : "code",
    "source" : [ "import org.apache.spark.sql.functions._\n", "import spark.implicits._\n", "\n", "val toDouble = udf[Double, String]( _.toDouble)" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "import org.apache.spark.sql.functions._\nimport spark.implicits._\ntoDouble: org.apache.spark.sql.expressions.UserDefinedFunction = UserDefinedFunction(<function1>,DoubleType,Some(List(StringType)))\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 9,
      "time" : "Took: 2.252s, at 2018-05-03 19:29"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "0F5690CD27E5490E870187D22C16F53F"
    },
    "cell_type" : "code",
    "source" : [ "val bowhead_liver = spark.readTSV(bowhead_liver_path, true).withColumn(\"liver\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n", "val gray_whale_liver = spark.readTSV(gray_whale_liver_path, true).withColumn(\"liver\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n", "val naked_mole_rat_liver = spark.readTSV(naked_mole_rat_liver_path, true).withColumn(\"liver\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n", "val minke_whale_liver = spark.readTSV( minke_whale_liver_path, true).withColumn(\"liver\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n", "\n", "val naked_mole_rat_ensembl_liver = spark.readTSV(naked_mole_rat_ensembl_liver_path, true).withColumn(\"liver\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n", "val cow_liver = spark.readTSV( cow_liver_path, true).withColumn(\"liver\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n", "val mouse_liver = spark.readTSV( mouse_liver_path, true).withColumn(\"liver\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n", "val human_liver = spark.readTSV( human_liver_path, true).withColumn(\"liver\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "bowhead_liver: org.apache.spark.sql.DataFrame = [transcript: string, Length: string ... 3 more fields]\ngray_whale_liver: org.apache.spark.sql.DataFrame = [transcript: string, Length: string ... 3 more fields]\nnaked_mole_rat_liver: org.apache.spark.sql.DataFrame = [transcript: string, Length: string ... 3 more fields]\nminke_whale_liver: org.apache.spark.sql.DataFrame = [transcript: string, Length: string ... 3 more fields]\nnaked_mole_rat_ensembl_liver: org.apache.spark.sql.DataFrame = [transcript: string, Length: string ... 3 more fields]\ncow_liver: org.apache.spark.sql.DataFrame = [transcript: string, Length: string ... 3 more fields]\nmouse_liver: org.apache.spark.sql.DataFrame = [transcript: string, Length: string ... 3 more fields]\nhuman_liver: org.apache.spark.sql.DataFrame = [transc..."
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 1,
      "time" : "Took: 24.783s, at 2018-05-03 19:38"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "80F84C022FB74CD49127A48E75FF72F3"
    },
    "cell_type" : "code",
    "source" : [ "val naked_mole_rat_kidney = spark.readTSV(naked_mole_rat_kidney_path, true).withColumn(\"kidney\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")  \n", "val bowhead_kidney = spark.readTSV(bowhead_kidney_path, true).withColumn(\"kidney\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n", "val gray_whale_kidney = spark.readTSV(gray_whale_kidney_path, true).withColumn(\"kidney\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n", "val minke_whale_kidney = spark.readTSV(minke_whale_kidney_path, true).withColumn(\"kidney\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n", "\n", "val naked_mole_rat_ensembl_kidney = spark.readTSV(naked_mole_rat_ensembl_kidney_path, true).withColumn(\"kidney\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n", "val cow_kidney = spark.readTSV( cow_kidney_path, true).withColumn(\"kidney\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n", "val mouse_kidney = spark.readTSV( mouse_kidney_path, true).withColumn(\"kidney\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n", "val human_kidney = spark.readTSV( human_kidney_path, true).withColumn(\"kidney\", toDouble($\"TPM\")).drop(\"TPM\").withColumnRenamed(\"Name\",\"transcript\")\n" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "naked_mole_rat_kidney: org.apache.spark.sql.DataFrame = [transcript: string, Length: string ... 3 more fields]\nbowhead_kidney: org.apache.spark.sql.DataFrame = [transcript: string, Length: string ... 3 more fields]\ngray_whale_kidney: org.apache.spark.sql.DataFrame = [transcript: string, Length: string ... 3 more fields]\nminke_whale_kidney: org.apache.spark.sql.DataFrame = [transcript: string, Length: string ... 3 more fields]\nnaked_mole_rat_ensembl_kidney: org.apache.spark.sql.DataFrame = [transcript: string, Length: string ... 3 more fields]\ncow_kidney: org.apache.spark.sql.DataFrame = [transcript: string, Length: string ... 3 more fields]\nmouse_kidney: org.apache.spark.sql.DataFrame = [transcript: string, Length: string ... 3 more fields]\nhuman_kidney: org.apache.spark.sql.DataFrame =..."
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 2,
      "time" : "Took: 6.262s, at 2018-05-03 19:40"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "BED1C73C5D5A45988F917BC99E05BB99"
    },
    "cell_type" : "code",
    "source" : [ "def join_liver_kidney(liver: DataFrame, kidney: DataFrame): DataFrame =\n", "    liver.join(kidney.withColumnRenamed(\"transcript\", \"kidney_transcript\"), $\"transcript\" === $\"kidney_transcript\").select(\"transcript\", \"liver\", \"kidney\")\n", "\n", "val gray_whale_expressions = join_liver_kidney(gray_whale_liver, gray_whale_kidney)\n", "val naked_mole_rat_expressions = join_liver_kidney(naked_mole_rat_liver, naked_mole_rat_kidney)\n", "val minke_whale_expressions = join_liver_kidney( minke_whale_liver,  minke_whale_kidney)\n", "val naked_mole_rat_ensembl_expressions = join_liver_kidney(naked_mole_rat_ensembl_liver, naked_mole_rat_ensembl_kidney)\n", "val mouse_expressions = join_liver_kidney(mouse_liver, mouse_kidney)\n", "val human_expressions = join_liver_kidney(human_liver, human_kidney)\n", "val cow_expressions = join_liver_kidney(cow_liver, cow_kidney)" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "join_liver_kidney: (liver: org.apache.spark.sql.DataFrame, kidney: org.apache.spark.sql.DataFrame)org.apache.spark.sql.DataFrame\ngray_whale_expressions: org.apache.spark.sql.DataFrame = [transcript: string, liver: double ... 1 more field]\nnaked_mole_rat_expressions: org.apache.spark.sql.DataFrame = [transcript: string, liver: double ... 1 more field]\nminke_whale_expressions: org.apache.spark.sql.DataFrame = [transcript: string, liver: double ... 1 more field]\nnaked_mole_rat_ensembl_expressions: org.apache.spark.sql.DataFrame = [transcript: string, liver: double ... 1 more field]\nmouse_expressions: org.apache.spark.sql.DataFrame = [transcript: string, liver: double ... 1 more field]\nhuman_expressions: org.apache.spark.sql.DataFrame = [transcript: string, liver: double ... 1 more field]\nc..."
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 3,
      "time" : "Took: 1.356s, at 2018-05-03 19:50"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "0ADE5A461AE04446854184F93A7855C5"
    },
    "cell_type" : "code",
    "source" : [ "/*\n", "gray_whale_expressions.coalesce(1).write.option(\"sep\",\"\\t\").option(\"header\",\"true\").csv(az(\"expressions/mixed/gray_whale_liver_kidney.tsv\"))\n", "naked_mole_rat_expressions.coalesce(1).write.option(\"sep\",\"\\t\").option(\"header\",\"true\").csv(az(\"expressions/mixed/naked_mole_rat_liver_kidney.tsv\"))\n", "minke_whale_expressions.coalesce(1).write.option(\"sep\",\"\\t\").option(\"header\",\"true\").csv(az(\"expressions/mixed/minke_whale_liver_kidney.tsv\"))\n", "*/" ],
    "outputs" : [ {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 13,
      "time" : "Took: 2.268s, at 2018-03-17 21:48"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "12A2D194C28546DEA20C48BCA785FA42"
    },
    "cell_type" : "code",
    "source" : [ " val exp_columns = List(\n", "    \"genage\",\n", "    \"transcript\",\n", "    \"liver\",\n", "    \"kidney\",\n", "    \"identity\",\n", "    \"aligment_length\",\n", "    \"mismatches\",\n", "    \"gaps\",\n", "    \"start_query\",\n", "    \"end_query\",\n", "    \"start_target\",\n", "    \"end_target\"\n", "  )" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "exp_columns: List[String] = List(genage, transcript, liver, kidney, identity, aligment_length, mismatches, gaps, start_query, end_query, start_target, end_target)\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 4,
      "time" : "Took: 0.899s, at 2018-05-03 19:50"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "8139D128DEA644109B8DF9EFA6C89DEA"
    },
    "cell_type" : "code",
    "source" : [ "val threshold = \"45\"\n", "val base = az(s\"lags/animals/${threshold}/blast6\")\n", "\n", "val gray_whale_path = base + \"/\" + \"animals_in_gray_whale.blast6\"\n", "val bowhead_alaska_liver_path = base + \"/\" + \"animals_in_bowhead_alaska_liver.blast6\"\n", "val bowhead_greenland_kidney_path = base + \"/\" + \"animals_in_bowhead_greenland_kidney.blast6\"\n", "val minke_whale_path = base + \"/\" + \"animals_in_minky_whale.blast6\"\n", "val naked_mole_rat_path = base + \"/\" +\"animals_in_naked_mole_rat.blast6\"\n", "val naked_mole_rat_ensembl_path = base + \"/\" + \"animals_in_naked_mole_rat_ensembl.blast6\"\n", "val mouse_path = base + \"/\" + \"animals_in_mouse.blast6\"\n", "val human_path = base + \"/\" + \"animals_in_human.blast6\"\n", "val cow_path = base + \"/\" + \"animals_in_cow.blast6\"" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "threshold: String = 45\nbase: String = wasbs://storage@pipelines1.blob.core.windows.net/lags/animals/45/blast6\ngray_whale_path: String = wasbs://storage@pipelines1.blob.core.windows.net/lags/animals/45/blast6/animals_in_gray_whale.blast6\nbowhead_alaska_liver_path: String = wasbs://storage@pipelines1.blob.core.windows.net/lags/animals/45/blast6/animals_in_bowhead_alaska_liver.blast6\nbowhead_greenland_kidney_path: String = wasbs://storage@pipelines1.blob.core.windows.net/lags/animals/45/blast6/animals_in_bowhead_greenland_kidney.blast6\nminke_whale_path: String = wasbs://storage@pipelines1.blob.core.windows.net/lags/animals/45/blast6/animals_in_minky_whale.blast6\nnaked_mole_rat_path: String = wasbs://storage@pipelines1.blob.core.windows.net/lags/animals/45/blast6/animals_in_naked_mole_rat.b..."
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 5,
      "time" : "Took: 0.930s, at 2018-05-03 20:28"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "7137D84FA9494DB48D36218736DE6067"
    },
    "cell_type" : "code",
    "source" : [ "val columns = List(\"genage\", \"transcript\", \"identity\", \"aligment_length\", \"mismatches\", \"gaps\", \"start_query\", \"end_query\", \"start_target\", \"end_target\") //last two non valid\n", "\n", "def withHeaders(df: DataFrame, cols: List[String], num:Int = 0): DataFrame = cols match {\n", "  case Nil => df\n", "  case head::tail =>   withHeaders(df.withColumnRenamed(\"_c\"+num, head), tail, num + 1)\n", "}" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "columns: List[String] = List(genage, transcript, identity, aligment_length, mismatches, gaps, start_query, end_query, start_target, end_target)\nwithHeaders: (df: org.apache.spark.sql.DataFrame, cols: List[String], num: Int)org.apache.spark.sql.DataFrame\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 6,
      "time" : "Took: 0.981s, at 2018-05-03 20:28"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "69C97BE5273342C1AA91A264D7007260"
    },
    "cell_type" : "code",
    "source" : [ "val headers = false\n", "val gray_whale_lags = withHeaders(spark.readTSV(gray_whale_path, headers), columns)\n", "val bowhead_alaska_lags = withHeaders(spark.readTSV(bowhead_alaska_liver_path, headers), columns)\n", "val bowhead_greenland_lags = withHeaders(spark.readTSV(bowhead_greenland_kidney_path, headers), columns)\n", "val minky_whale_lags = withHeaders(spark.readTSV(minke_whale_path, headers), columns)\n", "val naked_mole_rat_lags = withHeaders(spark.readTSV(naked_mole_rat_path, headers), columns)\n", "val naked_mole_rat_ensembl_lags = withHeaders(spark.readTSV(naked_mole_rat_ensembl_path, headers), columns) \n", "val mouse_lags = withHeaders(spark.readTSV(mouse_path, headers), columns)\n", "val human_lags = withHeaders(spark.readTSV(human_path, headers), columns)\n", "val cow_lags = withHeaders(spark.readTSV(cow_path, headers), columns)\n", "(gray_whale_lags.count, bowhead_alaska_lags.count, bowhead_greenland_lags.count, minky_whale_lags.count, naked_mole_rat_lags.count,\n", "naked_mole_rat_ensembl_lags.count, mouse_lags.count, human_lags.count, cow_lags.count\n", ")" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "headers: Boolean = false\ngray_whale_lags: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 10 more fields]\nbowhead_alaska_lags: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 10 more fields]\nbowhead_greenland_lags: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 10 more fields]\nminky_whale_lags: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 10 more fields]\nnaked_mole_rat_lags: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 10 more fields]\nnaked_mole_rat_ensembl_lags: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 10 more fields]\nmouse_lags: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 10 more fields]\nhuman_l..."
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : "(850,229,443,668,930,851,1051,1047,858)"
      },
      "output_type" : "execute_result",
      "execution_count" : 8,
      "time" : "Took: 5.158s, at 2018-05-03 20:30"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "D01AC49FE31D4FFC81C9EF4218E5330E"
    },
    "cell_type" : "code",
    "source" : [ "val gray_whale_lags_exp = gray_whale_lags.join(gray_whale_expressions, \"transcript\").select(exp_columns.head, exp_columns.tail:_*)\n", "val naked_mole_rat_lags_exp = naked_mole_rat_lags.join(naked_mole_rat_expressions, \"transcript\").select(exp_columns.head, exp_columns.tail:_*)\n", "val minke_whale_lags_exp = minky_whale_lags.join(minke_whale_expressions, \"transcript\").select(exp_columns.head, exp_columns.tail:_*)\n", "val naked_mole_rat_ensembl_lags_exp = naked_mole_rat_ensembl_lags.join(naked_mole_rat_ensembl_expressions, \"transcript\")\n", "  .select(exp_columns.head, exp_columns.tail:_*)\n", "val human_lags_exp = human_lags.join(human_expressions, \"transcript\").select(exp_columns.head, exp_columns.tail:_*)\n", "val mouse_lags_exp = mouse_lags.join(mouse_expressions, \"transcript\").select(exp_columns.head, exp_columns.tail:_*)\n", "val cow_lags_exp = cow_lags.join(cow_expressions, \"transcript\").select(exp_columns.head, exp_columns.tail:_*)\n", "\n", "(gray_whale_lags_exp.count, naked_mole_rat_lags_exp.count, minke_whale_lags_exp.count, \n", " naked_mole_rat_ensembl_lags_exp.count, human_lags_exp.count, mouse_lags_exp.count, cow_lags_exp.count)" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "gray_whale_lags_exp: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 10 more fields]\nnaked_mole_rat_lags_exp: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 10 more fields]\nminke_whale_lags_exp: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 10 more fields]\nnaked_mole_rat_ensembl_lags_exp: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 10 more fields]\nhuman_lags_exp: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 10 more fields]\nmouse_lags_exp: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 10 more fields]\ncow_lags_exp: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 10 more fields]\nres30: (Long, Long, Long, L..."
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : "(850,930,668,851,1047,1051,858)"
      },
      "output_type" : "execute_result",
      "execution_count" : 10,
      "time" : "Took: 10.593s, at 2018-05-03 20:35"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "166F496589A249AE81ACD777C73CE108"
    },
    "cell_type" : "code",
    "source" : [ "val rest = exp_columns.tail.tail.tail.tail\n", "val bowhead_liver_lags_exp = bowhead_liver.join(bowhead_alaska_lags,\"transcript\").select(exp_columns.head, \"transcript\"::\"liver\"::rest:_*)\n", "val bowhead_kidney_lags_exp = bowhead_kidney.join(bowhead_greenland_lags,\"transcript\").select(exp_columns.head, \"transcript\"::\"kidney\"::rest:_*)\n", "(bowhead_liver_lags_exp.count, bowhead_kidney_lags_exp.count)" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "rest: List[String] = List(identity, aligment_length, mismatches, gaps, start_query, end_query, start_target, end_target)\nbowhead_liver_lags_exp: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 9 more fields]\nbowhead_kidney_lags_exp: org.apache.spark.sql.DataFrame = [genage: string, transcript: string ... 9 more fields]\nres32: (Long, Long) = (229,443)\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : "(229,443)"
      },
      "output_type" : "execute_result",
      "execution_count" : 11,
      "time" : "Took: 3.628s, at 2018-05-03 20:35"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "8A649BDD41DC448BB5EB1575C551DD81"
    },
    "cell_type" : "code",
    "source" : [ " import spark.implicits._\n", "import org.apache.spark.sql.functions._\n", "import org.apache.spark.sql.ColumnName\n", "\n", "val genAgeExtract = udf[Double, String]((x: String) => x.replace(\">\", \"\").toDouble)\n", "\n", "val toLong = udf[Long, String]( _.toLong)\n", "\n", "def processAnimal(df: DataFrame) = {\n", "  df.na.fill(\"\")\n", "    .withColumn(\"liver_TPM\", toDouble($\"liver\")).drop(\"liver\")\n", "    .withColumn(\"kidney_TPM\", toDouble($\"kidney\")).drop(\"kidney\")\n", "    .withColumn(\"ident\", toDouble($\"identity\")).drop(\"identity\")\n", "    .withColumn(\"aligment_len\", toLong($\"aligment_length\")).drop(\"aligment_length\")\n", "    .withColumn(\"start_q\", toLong($\"start_query\")).drop(\"start_query\")\n", "    .withColumn(\"end_q\", toLong($\"end_query\")).drop(\"end_query\")\n", "    .withColumn(\"start_t\", toLong($\"start_target\")).drop(\"start_target\")\n", "    .withColumn(\"end_t\", toLong($\"end_target\")).drop(\"end_target\")\n", "    .withColumn(\"_tmp\", split($\"genage\", \"\\\\|\"))\n", "    .select(\n", "      genAgeExtract($\"_tmp\".getItem(0)).as(\"GenAge ID\"),\n", "      $\"transcript\",\n", "      $\"liver_TPM\",\n", "      $\"kidney_TPM\",\n", "      ($\"liver_TPM\" + $\"kidney_TPM\").as(\"avg_expression\"),\n", "      $\"ident\",\n", "      $\"aligment_len\",\n", "      $\"mismatches\",\n", "      $\"gaps\",\n", "      $\"start_q\",\n", "      $\"end_q\",\n", "      $\"start_t\",\n", "      $\"end_t\"\n", "  ).drop(\"_tmp\").drop($\"genage\")\n", "  .sort($\"avg_expression\".desc)\n", "}" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "import spark.implicits._\nimport org.apache.spark.sql.functions._\nimport org.apache.spark.sql.ColumnName\ngenAgeExtract: org.apache.spark.sql.expressions.UserDefinedFunction = UserDefinedFunction(<function1>,DoubleType,Some(List(StringType)))\ntoLong: org.apache.spark.sql.expressions.UserDefinedFunction = UserDefinedFunction(<function1>,LongType,Some(List(StringType)))\nprocessAnimal: (df: org.apache.spark.sql.DataFrame)org.apache.spark.sql.Dataset[org.apache.spark.sql.Row]\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 12,
      "time" : "Took: 2.371s, at 2018-05-03 20:35"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "7AB85678D24F48A8AF688BA8D8E49EBC"
    },
    "cell_type" : "code",
    "source" : [ "val results_columns = List(\n", "  \"GenAge ID\",\n", "  \"transcript\",\n", "  \"liver_TPM\",\n", "  \"kidney_TPM\",\n", "  \"avg_expression\",\n", "  \"ident\",\n", "  \"aligment_len\",\n", "  \"mismatches\",\n", "  \"gaps\",\n", "  \"start_q\",\n", "  \"end_q\",\n", "  \"start_t\",\n", "  \"end_t\"\n", ")\n", "val lags_models_columns = List(\"GenAge ID\", \"symbol\", \"name\", \"organism\", \"entrez gene id\", \"avg lifespan change (max obsv)\", \"lifespan effect\", \"longevity influence\")\n", "val lags_columns: List[String] = lags_models_columns ++ results_columns.tail\n" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "results_columns: List[String] = List(GenAge ID, transcript, liver_TPM, kidney_TPM, avg_expression, ident, aligment_len, mismatches, gaps, start_q, end_q, start_t, end_t)\nlags_models_columns: List[String] = List(GenAge ID, symbol, name, organism, entrez gene id, avg lifespan change (max obsv), lifespan effect, longevity influence)\nlags_columns: List[String] = List(GenAge ID, symbol, name, organism, entrez gene id, avg lifespan change (max obsv), lifespan effect, longevity influence, transcript, liver_TPM, kidney_TPM, avg_expression, ident, aligment_len, mismatches, gaps, start_q, end_q, start_t, end_t)\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 13,
      "time" : "Took: 2.097s, at 2018-05-03 20:35"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "ADD02882B5124EEAA8FF9D6D830B942D"
    },
    "cell_type" : "code",
    "source" : [ "val lags_models_columns = List(\"GenAge ID\", \"symbol\", \"name\", \"organism\", \"entrez gene id\", \"avg lifespan change (max obsv)\", \"lifespan effect\", \"longevity influence\")\n", "val lags_columns: List[String] = lags_models_columns ++ results_columns.tail\n", "\n", "val lags_models = spark.readTSV(az(\"lags/genage_models.tsv\"), true)\n", "def join_with_models(df: DataFrame) = {\n", "  lags_models.join(df, \"GenAge ID\").select(lags_columns.head, lags_columns.tail:_*)\n", "}\n", "lags_models" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "lags_models_columns: List[String] = List(GenAge ID, symbol, name, organism, entrez gene id, avg lifespan change (max obsv), lifespan effect, longevity influence)\nlags_columns: List[String] = List(GenAge ID, symbol, name, organism, entrez gene id, avg lifespan change (max obsv), lifespan effect, longevity influence, transcript, liver_TPM, kidney_TPM, avg_expression, ident, aligment_len, mismatches, gaps, start_q, end_q, start_t, end_t)\nlags_models: org.apache.spark.sql.DataFrame = [GenAge ID: string, symbol: string ... 6 more fields]\njoin_with_models: (df: org.apache.spark.sql.DataFrame)org.apache.spark.sql.DataFrame\nres36: org.apache.spark.sql.DataFrame = [GenAge ID: string, symbol: string ... 6 more fields]\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : "<div class=\"df-canvas\">\n      <script data-this=\"{&quot;dataId&quot;:&quot;anon8aa16de4a061c91dd119511664ef2b75&quot;,&quot;partitionIndexId&quot;:&quot;anon440b9826972d79c662a84f59ed9a8aa4&quot;,&quot;numPartitions&quot;:87,&quot;dfSchema&quot;:{&quot;type&quot;:&quot;struct&quot;,&quot;fields&quot;:[{&quot;name&quot;:&quot;GenAge ID&quot;,&quot;type&quot;:&quot;string&quot;,&quot;nullable&quot;:true,&quot;metadata&quot;:{}},{&quot;name&quot;:&quot;symbol&quot;,&quot;type&quot;:&quot;string&quot;,&quot;nullable&quot;:true,&quot;metadata&quot;:{}},{&quot;name&quot;:&quot;name&quot;,&quot;type&quot;:&quot;string&quot;,&quot;nullable&quot;:true,&quot;metadata&quot;:{}},{&quot;name&quot;:&quot;organism&quot;,&quot;type&quot;:&quot;string&quot;,&quot;nullable&quot;:true,&quot;metadata&quot;:{}},{&quot;name&quot;:&quot;entrez gene id&quot;,&quot;type&quot;:&quot;string&quot;,&quot;nullable&quot;:true,&quot;metadata&quot;:{}},{&quot;name&quot;:&quot;avg lifespan change (max obsv)&quot;,&quot;type&quot;:&quot;string&quot;,&quot;nullable&quot;:true,&quot;metadata&quot;:{}},{&quot;name&quot;:&quot;lifespan effect&quot;,&quot;type&quot;:&quot;string&quot;,&quot;nullable&quot;:true,&quot;metadata&quot;:{}},{&quot;name&quot;:&quot;longevity influence&quot;,&quot;type&quot;:&quot;string&quot;,&quot;nullable&quot;:true,&quot;metadata&quot;:{}}]}}\" type=\"text/x-scoped-javascript\">/*<![CDATA[*/req(['../javascripts/notebook/dataframe','../javascripts/notebook/consoleDir'], \n      function(dataframe, extension) {\n        dataframe.call(data, this, extension);\n      }\n    );/*]]>*/</script>\n      <link rel=\"stylesheet\" href=\"/assets/stylesheets/ipython/css/dataframe.css\" type=\"text/css\"/>\n    </div>"
      },
      "output_type" : "execute_result",
      "execution_count" : 14,
      "time" : "Took: 2.770s, at 2018-05-03 20:35"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "presentation" : {
        "tabs_state" : "{\n  \"tab_id\": \"#tab1046884362-0\"\n}",
        "pivot_chart_state" : "{\n  \"hiddenAttributes\": [],\n  \"menuLimit\": 200,\n  \"cols\": [],\n  \"rows\": [],\n  \"vals\": [],\n  \"exclusions\": {},\n  \"inclusions\": {},\n  \"unusedAttrsVertical\": 85,\n  \"autoSortUnusedAttrs\": false,\n  \"inclusionsInfo\": {},\n  \"aggregatorName\": \"Count\",\n  \"rendererName\": \"Table\"\n}"
      },
      "id" : "70FF47EF75824362823A2297A0D49492"
    },
    "cell_type" : "code",
    "source" : [ "def process_with_models(df: DataFrame) = join_with_models( processAnimal(df))\n", "//\"1978|ENSMUST00000173997.1|Rbm38|ENSMUSG00000027510.17|56190|Mus_musculus|22.0|Decrease|Pro-Longevity|chr2|173022437|173033709|FORWARD\"" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "process_with_models: (df: org.apache.spark.sql.DataFrame)org.apache.spark.sql.DataFrame\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 15,
      "time" : "Took: 1.024s, at 2018-05-03 20:36"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "89765E236A414E0C84D1DBD16D442991"
    },
    "cell_type" : "code",
    "source" : [ "val fields =  spark.readTSV(az(s\"lags/genage_models_export_full.tsv\"), true)\n", "///storage/lags" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "fields: org.apache.spark.sql.DataFrame = [Gene ID: string, Entrez Gene ID: string ... 14 more fields]\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 16,
      "time" : "Took: 1.482s, at 2018-05-03 20:36"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "presentation" : {
        "tabs_state" : "{\n  \"tab_id\": \"#tab1731311106-0\"\n}",
        "pivot_chart_state" : "{\n  \"hiddenAttributes\": [],\n  \"menuLimit\": 200,\n  \"cols\": [],\n  \"rows\": [],\n  \"vals\": [],\n  \"exclusions\": {},\n  \"inclusions\": {},\n  \"unusedAttrsVertical\": 85,\n  \"autoSortUnusedAttrs\": false,\n  \"inclusionsInfo\": {},\n  \"aggregatorName\": \"Count\",\n  \"rendererName\": \"Table\"\n}"
      },
      "id" : "2549A4AD255A4ED093394A0C9BB7B84C"
    },
    "cell_type" : "code",
    "source" : [ "val extra = fields.select(\"Gene ID\",\"Unigene ID\",\"Ensembl ID\",\"Alias\", \"Phenotype Description\", \"Max Lifespan Change\", \"Method\", \"Bibliographic reference\")\n", "\n" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "extra: org.apache.spark.sql.DataFrame = [Gene ID: string, Unigene ID: string ... 6 more fields]\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 17,
      "time" : "Took: 0.998s, at 2018-05-03 20:36"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "760121BBEC4A4E04B5DA983001800217"
    },
    "cell_type" : "code",
    "source" : [ "val gray_whale_animal_lags = process_with_models(gray_whale_lags_exp).join(extra, $\"Gene ID\"===$\"GenAge ID\").drop($\"Gene ID\")\n", "val minky_whale_animal_lags = process_with_models(minke_whale_lags_exp).join(extra, $\"Gene ID\"===$\"GenAge ID\").drop($\"Gene ID\")\n", "val naked_mole_rat_animal_lags = process_with_models(naked_mole_rat_lags_exp).join(extra, $\"Gene ID\"===$\"GenAge ID\").drop($\"Gene ID\")\n", "val naked_mole_rat_ensembl_animal_lags = process_with_models(naked_mole_rat_ensembl_lags_exp).join(extra, $\"Gene ID\"===$\"GenAge ID\").drop($\"Gene ID\")\n", "val human_animal_lags = process_with_models(human_lags_exp).join(extra, $\"Gene ID\"===$\"GenAge ID\").drop($\"Gene ID\")\n", "val mouse_animal_lags = process_with_models(mouse_lags_exp).join(extra, $\"Gene ID\"===$\"GenAge ID\").drop($\"Gene ID\")\n", "val cow_animal_lags = process_with_models(cow_lags_exp).join(extra, $\"Gene ID\"===$\"GenAge ID\").drop($\"Gene ID\")\n" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "gray_whale_animal_lags: org.apache.spark.sql.DataFrame = [GenAge ID: string, symbol: string ... 25 more fields]\nminky_whale_animal_lags: org.apache.spark.sql.DataFrame = [GenAge ID: string, symbol: string ... 25 more fields]\nnaked_mole_rat_animal_lags: org.apache.spark.sql.DataFrame = [GenAge ID: string, symbol: string ... 25 more fields]\nnaked_mole_rat_ensembl_animal_lags: org.apache.spark.sql.DataFrame = [GenAge ID: string, symbol: string ... 25 more fields]\nhuman_animal_lags: org.apache.spark.sql.DataFrame = [GenAge ID: string, symbol: string ... 25 more fields]\nmouse_animal_lags: org.apache.spark.sql.DataFrame = [GenAge ID: string, symbol: string ... 25 more fields]\ncow_animal_lags: org.apache.spark.sql.DataFrame = [GenAge ID: string, symbol: string ... 25 more fields]\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 18,
      "time" : "Took: 2.984s, at 2018-05-03 20:37"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "presentation" : {
        "tabs_state" : "{\n  \"tab_id\": \"#tab1335433788-0\"\n}",
        "pivot_chart_state" : "{\n  \"hiddenAttributes\": [],\n  \"menuLimit\": 200,\n  \"cols\": [],\n  \"rows\": [],\n  \"vals\": [],\n  \"exclusions\": {},\n  \"inclusions\": {},\n  \"unusedAttrsVertical\": 85,\n  \"autoSortUnusedAttrs\": false,\n  \"inclusionsInfo\": {},\n  \"aggregatorName\": \"Count\",\n  \"rendererName\": \"Table\"\n}"
      },
      "id" : "0DB8C2F646C746F88846A38F81EDD227"
    },
    "cell_type" : "code",
    "source" : [ "gray_whale_animal_lags.coalesce(1).write.option(\"sep\",\"\\t\").option(\"header\",\"true\").csv(az(s\"lags/animals/${threshold}/lags/gray_whale_animal_lags.tsv\"))\n", "minky_whale_animal_lags.coalesce(1).write.option(\"sep\",\"\\t\").option(\"header\",\"true\").csv(az(s\"lags/animals/${threshold}/lags/minky_whale_animal_lags.tsv\"))\n", "naked_mole_rat_animal_lags.coalesce(1).write.option(\"sep\",\"\\t\").option(\"header\",\"true\").csv(az(s\"lags/animals/${threshold}/lags/naked_mole_rat_animal_lags.tsv\"))\n" ],
    "outputs" : [ {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 41,
      "time" : "Took: 13.269s, at 2018-03-27 23:47"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "C7DD864BBD98443086EFCF499EB2CA73"
    },
    "cell_type" : "code",
    "source" : [ "naked_mole_rat_ensembl_animal_lags.coalesce(1).write.option(\"sep\",\"\\t\").option(\"header\",\"true\")\n", ".csv(az(s\"lags/animals/${threshold}/lags/naked_mole_rat_ensembl_animal_lags.tsv\"))\n", "human_animal_lags.coalesce(1).write.option(\"sep\",\"\\t\").option(\"header\",\"true\")\n", ".csv(az(s\"lags/animals/${threshold}/lags/human_animal_lags.tsv\"))\n", "mouse_animal_lags.coalesce(1).write.option(\"sep\",\"\\t\").option(\"header\",\"true\")\n", ".csv(az(s\"lags/animals/${threshold}/lags/mouse_animal_lags.tsv\"))\n", "cow_animal_lags.coalesce(1).write.option(\"sep\",\"\\t\").option(\"header\",\"true\")\n", ".csv(az(s\"lags/animals/${threshold}/lags/cow_animal_lags.tsv\"))\n" ],
    "outputs" : [ {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 19,
      "time" : "Took: 18.314s, at 2018-05-03 20:39"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "632840A794F340539571E1D8058DFF9A"
    },
    "cell_type" : "code",
    "source" : [ "def processLiver(df: DataFrame) = {\n", "  df.na.fill(\"\")\n", "    .withColumn(\"liver_TPM\", toDouble($\"liver\")).drop(\"liver\")\n", "    .withColumn(\"ident\", toDouble($\"identity\")).drop(\"identity\")\n", "    .withColumn(\"aligment_len\", toLong($\"aligment_length\")).drop(\"aligment_length\")\n", "    .withColumn(\"start_q\", toLong($\"start_query\")).drop(\"start_query\")\n", "    .withColumn(\"end_q\", toLong($\"end_query\")).drop(\"end_query\")\n", "    .withColumn(\"start_t\", toLong($\"start_target\")).drop(\"start_target\")\n", "    .withColumn(\"end_t\", toLong($\"end_target\")).drop(\"end_target\")\n", "    .withColumn(\"_tmp\", split($\"genage\", \"\\\\|\"))\n", "    .select(\n", "      genAgeExtract($\"_tmp\".getItem(0)).as(\"GenAge ID\"),\n", "      $\"transcript\",\n", "      $\"liver_TPM\",\n", "      $\"ident\",\n", "      $\"aligment_len\",\n", "      $\"mismatches\",\n", "      $\"gaps\",\n", "      $\"start_q\",\n", "      $\"end_q\",\n", "      $\"start_t\",\n", "      $\"end_t\"\n", "    ).drop(\"_tmp\").drop($\"genage\")\n", "    .sort($\"liver_TPM\".desc)\n", "}\n", "\n", "def processKidney(df: DataFrame) = {\n", "  df.na.fill(\"\")\n", "    .withColumn(\"kidney_TPM\", toDouble($\"kidney\")).drop(\"kidney\")\n", "    .withColumn(\"ident\", toDouble($\"identity\")).drop(\"identity\")\n", "    .withColumn(\"aligment_len\", toLong($\"aligment_length\")).drop(\"aligment_length\")\n", "    .withColumn(\"start_q\", toLong($\"start_query\")).drop(\"start_query\")\n", "    .withColumn(\"end_q\", toLong($\"end_query\")).drop(\"end_query\")\n", "    .withColumn(\"start_t\", toLong($\"start_target\")).drop(\"start_target\")\n", "    .withColumn(\"end_t\", toLong($\"end_target\")).drop(\"end_target\")\n", "    .withColumn(\"_tmp\", split($\"genage\", \"\\\\|\"))\n", "    .select(\n", "      genAgeExtract($\"_tmp\".getItem(0)).as(\"GenAge ID\"),\n", "      $\"transcript\",\n", "      $\"kidney_TPM\",\n", "      $\"ident\",\n", "      $\"aligment_len\",\n", "      $\"mismatches\",\n", "      $\"gaps\",\n", "      $\"start_q\",\n", "      $\"end_q\",\n", "      $\"start_t\",\n", "      $\"end_t\"\n", "    ).drop(\"_tmp\").drop($\"genage\")\n", "    .sort($\"kidney_TPM\".desc)\n", "}\n", "\n", "val lags_liver_columns: List[String] = lags_models_columns ++ results_columns.tail.filterNot(s=>s.contains(\"kidney\") || s.contains(\"avg_expression\"))\n", "val lags_kidney_columns: List[String] = lags_models_columns ++ results_columns.tail.filterNot(s=>s.contains(\"liver\") || s.contains(\"avg_expression\"))\n", "def join_liver_with_models(df: DataFrame) = {\n", "  lags_models.join(df, \"GenAge ID\").select(lags_liver_columns.head, lags_liver_columns.tail:_*)\n", "}\n", "\n", "def join_kidney_with_models(df: DataFrame) = {\n", "  lags_models.join(df, \"GenAge ID\").select(lags_kidney_columns.head, lags_kidney_columns.tail:_*)\n", "}\n", "\n", "def process_with_liver_models(df: DataFrame) = join_liver_with_models( processLiver(df))\n", "def process_with_kidney_models(df: DataFrame) = join_kidney_with_models( processKidney(df))\n" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "processLiver: (df: org.apache.spark.sql.DataFrame)org.apache.spark.sql.Dataset[org.apache.spark.sql.Row]\nprocessKidney: (df: org.apache.spark.sql.DataFrame)org.apache.spark.sql.Dataset[org.apache.spark.sql.Row]\nlags_liver_columns: List[String] = List(GenAge ID, symbol, name, organism, entrez gene id, avg lifespan change (max obsv), lifespan effect, longevity influence, transcript, liver_TPM, ident, aligment_len, mismatches, gaps, start_q, end_q, start_t, end_t)\nlags_kidney_columns: List[String] = List(GenAge ID, symbol, name, organism, entrez gene id, avg lifespan change (max obsv), lifespan effect, longevity influence, transcript, kidney_TPM, ident, aligment_len, mismatches, gaps, start_q, end_q, start_t, end_t)\njoin_liver_with_models: (df: org.apache.spark.sql.DataFrame)org.apache.spa..."
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 20,
      "time" : "Took: 1.516s, at 2018-05-03 20:39"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "23239B42D2EE4ECFBC15FF6B61451DEE"
    },
    "cell_type" : "code",
    "source" : [ "val bowhead_liver_animal_lags = process_with_liver_models(bowhead_liver_lags_exp).join(extra, $\"Gene ID\"===$\"GenAge ID\").drop($\"Gene ID\")\n", "val bowhead_kidney_animal_lags = process_with_kidney_models(bowhead_kidney_lags_exp).join(extra, $\"Gene ID\"===$\"GenAge ID\").drop($\"Gene ID\")" ],
    "outputs" : [ {
      "name" : "stdout",
      "output_type" : "stream",
      "text" : "bowhead_liver_animal_lags: org.apache.spark.sql.DataFrame = [GenAge ID: string, symbol: string ... 23 more fields]\nbowhead_kidney_animal_lags: org.apache.spark.sql.DataFrame = [GenAge ID: string, symbol: string ... 23 more fields]\n"
    }, {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 21,
      "time" : "Took: 1.673s, at 2018-05-03 20:41"
    } ]
  }, {
    "metadata" : {
      "trusted" : true,
      "input_collapsed" : false,
      "collapsed" : false,
      "id" : "AA30C24EA0FC4953A701C467C451D675"
    },
    "cell_type" : "code",
    "source" : [ "bowhead_liver_animal_lags.coalesce(1).write.option(\"sep\",\"\\t\").option(\"header\",\"true\").csv(az(s\"lags/animals/${threshold}/lags/bowhead_liver_animal_lags.tsv\"))\n", "bowhead_kidney_animal_lags.coalesce(1).write.option(\"sep\",\"\\t\").option(\"header\",\"true\").csv(az(s\"lags/animals/${threshold}/lags/bowhead_kidney_animal_lags.tsv\"))\n" ],
    "outputs" : [ {
      "metadata" : { },
      "data" : {
        "text/html" : ""
      },
      "output_type" : "execute_result",
      "execution_count" : 43,
      "time" : "Took: 10.626s, at 2018-03-27 23:47"
    } ]
  } ],
  "nbformat" : 4
}